src/runtime/pkg/sev/sev.go

```diff
@@ -163,3 +163,47 @@ func CalculateLaunchDigest(firmwarePath, kernelPath, initrdPath, cmdline string)
 	copy(res[:], digest.Sum(nil))
 	return res, nil
 }
+
+// CalculateSEVESLaunchDigest returns the sha256 encoded SEV-ES launch digest
+// based off the current firmware, kernel, initrd, and the kernel cmdline, and
+// the number of vcpus and their type
+func CalculateSEVESLaunchDigest(vcpus int, vcpuSig VCPUSig, firmwarePath, kernelPath, initrdPath, cmdline string) (res [sha256.Size]byte, err error) {
+	f, err := os.Open(firmwarePath)
+	if err != nil {
+		return res, err
+	}
+	defer f.Close()
+
+	digest := sha256.New()
+	if _, err := io.Copy(digest, f); err != nil {
+		return res, err
+	}
+
+	if kernelPath != "" {
```

Comment:
Is this the right condition? It seems like the `kernelPath` will always exist in practice. Do we need to support cases without the kernel hashes? Currently the kernel hashes QEMU parameter is set when the guest pre-attestation flag is on. When guest pre-attestation is off, we don't connect to the KBS. In other words, when we don't have the kernel hashes set, we also don't give the expected measurement to anyone because we don't validate the measurement. So maybe we don't need this check at all? Still, feels like maybe the option to do it without kernel hashes should exist.

Reply:
Indeed inside kata `kernelPath` is always set. I introduced this condition in #5337 so that the `sev.calculateLaunchDigest()` function can be used also outside the kata world, as a standalone go package. Something like:

```
package main

import "fmt"
import "github.com/kata-containers/kata-containers/src/runtime/pkg/sev"

func main() {
        ld, err := sev.CalculateLaunchDigest("/path/to/OVMF.fd", "", "", "")
        fmt.Println("err=", err, "ld=", ld)
}
```

This allows writing standalone Go programs which calculate the expected digests.

@fitzthum Do you think it's worth adding a comment explaining this above the `kernelPath` condition?

----

Tip for people who want to try this at home: the `runtime/pkg/sev` package doesn't exist in branch `main`. If you want to use the above code you'll first need to `go get github.com/kata-containers/kata-containers/src/runtime/pkg/sev@CCv0` (note the branch name at the end).

Reply:
Ok, that makes sense. A comment might be handy.